merge/update.go

```diff
@@ -250,31 +250,47 @@ func (s *Updater) addBackOwnedItems(merged, pruned *typed.TypedValue, managedFie
 		}
 		managedAtVersion[managerSet.APIVersion()] = managedAtVersion[managerSet.APIVersion()].Union(managerSet.Set())
 	}
-	for version, managed := range managedAtVersion {
-		merged, err = s.Converter.Convert(merged, version)
-		if err != nil {
-			if s.Converter.IsMissingVersionError(err) {
+	// We need a retry loop here to ensure that when we convert the pruned object
+	// to a specific version, it already has all the required fields back.
+	for {
```

Comment:
In the very worst case, where there's exactly one conversion sequence that works, this retry loop's runtime complexity appears to be N! for N versions -- I think we can't afford that even though the average case is not that bad.

I *do* think we can afford to sort by version name so that it's at least deterministic.

I guess we don't hit this problem on built in types because we know how to default them? This is the part I'm still a bit confused about.

If determinism + asking conversion webhooks to first default isn't enough, then we need to cap the number of tries this loop can make, and ideally-- cache the successful order so that we only have to figure it out once. But I hope we don't have to do that.

Reply:
About the complexity, I think the worst case is O(N^2) rather than N!. The retry loop does not traverse the permutation of versions, instead, it does at least one successful conversion in each iteration (will fail fast if not) and then removes that converted version and does the next iteration. So the actual iteration count for the worst case would be 1+2+3+...+N which is N(N+1)/2.

It's hard to imagine there would be a resource type with more than 10 versions, let alone more. So it seems that the solution's performance is not that bad?

Reply:
OK, you're right about the complexity, sorry, I missed where it removes the version on success. 

However, I think that doing that leaves cases that fail. Suppose there are four versions, and it's safe to convert in the order A->B->C->D, but most other conversions fail. Suppose the first order we try is B->C->A->D. B->C is successful, triggering removal of C, but now we're left with versions A, B, and D, and there's no possible order which will be successful (you can't get from B to D, nor can you get from D to A). So, the proposed fix only solves a subset of path-dependent conversion cases.

The current assumption enforced on built-ins is that we default before converting, and conversions must not ever fail (i.e., they are never path-dependent). I'm not sure we should relax this constraint at all.

Reply:
I'd like to clarify that the root cause of this failure is not the conversion order but the lack of **required** fields in a specific version. As long as an object is integral (each required field has a value) at a specific version, it can be converted to any other version successfully. So the conversion failure here is not really caused by a wrong order but by an imcomplete object (which is the **pruned** object) when converting. And what the fix do is not finding some order but making the pruned object integral at its current version (by adding back managed fields at this version).

Now we can return to the ABCD example, the problem is not caused by the A->B->C->D order (actually there's no conversion order limit) but by: there's a required non-empty field `fa` in A, and similarly `fb`, `fc`, `fd`. Now a pruned object comes to our loop, say it's at version A with `fa` pruned, and the loop order is B->C->A->D, here's what'll happen:
1. try convert to B, C, both fail because A is not integral (lack required field `fa`)
2. convert to A, success because there's actually no convert
3. add back owned items at version A (which include the required field `fa`), now the pruned obj is integral at version A and is ready to convert to any other version
4. convert to D, success because obj at version A is integral
5. add back owned items at version D (which include the required field `fd`), now the pruned obj is also integral at version D
6. then every following conversion will success because the obj will always be integral at its current version from now on

This bug is indeed a little tricky that it misled both of us lol. After think it twice and list the above step by step, I found I'm also wrong at the complexity, it's not O(N^2) but O(N)! The fix is not finding an order but to make sure the pruned obj is integral at first version (the version where the `pruned` parameter of the `addBackOwnedItems` function at) before it's converted to any other version. So I found an even better solution that we just get the version of `pruned` then do the add back logic on that version first and everything will go smoothly then. This makes the complexity of the fix become O(1). wdyt @lavalamp

Reply:
Thanks for the great analysis! Your described fix sounds great, I can't object to an O(1) fix ;)